### PR TITLE
Now compatible with Philipp's AnnData class

### PR DESCRIPTION
**Pull request by falexwolf**

### CODE REVIEW

File: scanpy/ann_data.py [hunk 1]
```diff
@@ -91,21 +92,20 @@ def _check_dimensions(data, smp, var):
...
-        u"""
-        Annotated Matrix
+    def __init__(self, ddata=None, X=None, smp=None, var=None, vis=None, **meta):
+        """
+        Annotated Data
 
-        Represents a sample × variable matrix (e.g. cell × gene) with the
-        possibility to store an arbitrary number of annotations for both
-        samples and variables.
+        Stores a data matrix X of dimensions nr_samples x nr_variables,
```
Comment (flying-sheep) [inline scanpy/ann_data.py:99]: why did you remove the “×” symbols? isn’t it clearer that way? i always need a moment to process that some “x” isn’t actually an x, but an ASCII approximation to “×”

File: scanpy/ann_data.py [hunk 2]
```diff
@@ -164,27 +202,37 @@ def __init__(self, X, smp=None, var=None, vis=None, **meta):
...
+                        self.smp[k] = v
+                elif key == 'col':
+                    for k, v in ddata['col'].items():
+                        self.var[k] = v
+                elif key == 'var':
+                    for k, v in ddata['var'].items():
+                        self.var[k] = v
+                    
+    def smp_keys(self):
+        return list(self.smp.dtype.names)[1:]
```
Comment (flying-sheep) [inline scanpy/ann_data.py:223]: `smp_names` aren’t really guarenteed to be first, right?

either this should be changed to `return set(self.smp.dtype.names) - {SMP_NAMES}` or we should make sure to reorder the columns appropriately when setting them

i also think those should be properties instead of functions and maybe named “smp_columns”. “keys” is dictionary terminology and we have a structured array.

File: scanpy/ann_data.py [hunk 3]
```diff
@@ -164,27 +202,37 @@ def __init__(self, X, smp=None, var=None, vis=None, **meta):
...
+
+    def var_keys(self):
+        return list(self.var.dtype.names)[1:]
+
+    def to_dict(self):
+        smp = {k: self.smp[k] for k in self.smp_keys() if not k=='smp_names'}
+        var = {k: self.var[k] for k in self.var_keys() if not k=='var_names'}
```
Comment (flying-sheep) [inline scanpy/ann_data.py:230]: please use the constants, it’s less error prone.

also, smp_names/var_names are already excluded in that method.

File: scanpy/ann_data.py (hunk 1 above)
Comment (falexwolf) [reply scanpy/ann_data.py:99]: Because of compatibility with Python 2, which threw an error. I think that could have been fixed by specifying the encoding. But then, everything should be done the same way. I can restore it when doing some final polishing on the docstrings.

File: scanpy/ann_data.py (hunk 2 above)
Comment (falexwolf) [reply scanpy/ann_data.py:223]: yes, that's a hack. but I think the np.recarray is ordered, right?

both your suggestions are fine. i just wanted it to get to work. I'm happy if you go over it again. maybe not right now, as everything is tested. but we can sit together maybe tomorrow, if you want.

File: scanpy/ann_data.py (hunk 1 above)
Comment (flying-sheep) [reply scanpy/ann_data.py:99]: yeah: `# coding: utf-8` and an `u` in front fo the docstring.

> But then, everything should be done the same way

i don’t understand what you mean by that. everything *is* done the same way. the default encoding in python 2 is ASCII, which is a subset of utf-8, which means that having or not having the UTF-8 encoding declaration is irrelevant as long as there is no error immediately when loading the module.

File: scanpy/ann_data.py (hunk 3 above)
Comment (falexwolf) [reply scanpy/ann_data.py:230]: yes, you're completely right. the whole thing needs to be redone. i haven't slept a minute for 29 hours. ;)

**Comment by flying-sheep** [comment]
great stuff 😄

File: scanpy/ann_data.py (hunk 1 above)
Comment (falexwolf) [reply scanpy/ann_data.py:99]: ok! I'll restore it then.

**Comment by falexwolf** [comment]
no, not great, but it work's and one should now be much better settled for the future with AnnData.

for example, the gene plots and different subgroups work. if you have a good suggestion for a default color map for continuous and categorial columns in smp, I'm very happy to adapt it. :)

https://github.com/falexwolf/collab_alex/blob/master/scanpy/examples/maehr17.md

or here directly in the main readme

https://github.com/theislab/scanpy#moignard15

File: scanpy/ann_data.py (hunk 2 above)
Comment (flying-sheep) [reply scanpy/ann_data.py:223]: it’s ordered, but doing `AnnData(X, dict(a=..., smp_names=...))` is valid and will* result in smp_names being the second column.


*on python ≤3.5 randomly, on python 3.6 always

File: scanpy/ann_data.py (hunk 3 above)
Comment (flying-sheep) [reply scanpy/ann_data.py:230]: why would you do that, go to sleep!
